csrc/scheduler/normalization.cpp

```diff
@@ -225,13 +225,11 @@ std::shared_ptr<ReductionParams> innerOuterPersistentHeuristic(
     iop.gdimy = blocks_per_sm * device_multiprocessor_count;
 
     // Step-3, OuterParams, Iteration dim: vectorization_factor_outer(reuse),
-    // bdimy, gdimy (in previous step). vectorization_factor_outer is set to 2
-    // as a small workload per thread is preferred for small sizes and we only
-    // process vectorized cases.
-    iop.bdimy = std::min(
-        ceilDiv(inner_dim_numel / iop.vectorization_factor_outer, iop.gdimy),
-        scheduler_utils::safeDiv(threads_per_block_mrpb, iop.bdimx));
-    iop.bdimy = iop.bdimy;
+    // bdimy, gdimy (in previous step). Here bdimy is set to the largest
+    // possible value limited by threads_per_block_mrpb to maximize
+    // multi-inner-reductions per block. It may be larger than the value
+    // required by the workload in iteration dim of the outer reduction.
```

Comment:
It's still unclear to me why making `bdimy` larger than necessary can improve a performance. Could you show some example case?

Reply:
"larger than necessary" is only for the 2nd part of the outer reduction. 
It increased the number of inner reductions per block.
So the performance gain is mainly from the inner reduction part and without penalty for the outer reduction part.
e.g. previously there are only 16 threads per CTA for case with hidden size of 128. With revision, it changed to 512 and the performance increased from 43 GB/s to 131 GB/s.
The parameters of other cases are also listed in the 1st comment of this PR.

Reply:
So, in the case of the inner reduction, bdimy is used for the iteration domain. gdimy is also used for the iteration domain. Don't we also want to check if `bdimy * gdimy` is not too large compared to the iteration size?

Reply:
Good point. I didn't check what if we avoid wasting resources along iteration domain for inner reduction. I added another goup of benchmarks with a small batch size of 128. The test shows adding 
```
    int64_t bdimy_max = std::max(
        ceilDiv(outer_dim_numel, iop.gdimy),
        ceilDiv(inner_dim_numel, iop.vectorization_factor_outer * iop.gdimy));
```
can slightly increase the performance. It allows us more uniformly distribute the small workload to all SMs instead of trying to fully utilize several SMs. e.g. for case 128 x 128, without this revision the coding is doing 32 reductions per block and only 4 blocks are active. With this change, the code is doing 2 reductions per block and 64 blocks are active.
Without this revision:
```
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/128/manual_time         37.7 us          121 us          371 bytes_per_second=2.45499G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/32/16)/grid(1/108/1)/2048]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/256/manual_time         32.5 us          116 us          431 bytes_per_second=5.67479G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/16/32)/grid(1/108/1)/2048]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/384/manual_time         24.5 us          107 us          571 bytes_per_second=11.2914G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/10/48)/grid(1/108/1)/1920]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/512/manual_time         24.3 us          107 us          576 bytes_per_second=15.1727G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/8/64)/grid(1/108/1)/2048]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/640/manual_time         23.1 us          105 us          604 bytes_per_second=19.9513G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/6/80)/grid(1/108/1)/1920]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/768/manual_time         20.8 us          103 us          674 bytes_per_second=26.5711G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/5/96)/grid(1/108/1)/1920]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/896/manual_time         26.9 us          109 us          523 bytes_per_second=23.9694G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/4/112)/grid(1/108/1)/1792]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/1023/manual_time        24.8 us          105 us          562 bytes_per_second=29.6248G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 4 / /Launch_Parameters[block(1/2/256)/grid(1/216/1)/2048]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/1024/manual_time        24.1 us          106 us          580 bytes_per_second=30.5341G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/4/128)/grid(1/108/1)/2048]
```
Limit bdimy
```
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/128/manual_time         21.8 us          103 us          646 bytes_per_second=4.25199G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/2/16)/grid(1/108/1)/128]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/256/manual_time         19.4 us          101 us          717 bytes_per_second=9.49036G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/2/32)/grid(1/108/1)/256]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/384/manual_time         20.8 us          103 us          672 bytes_per_second=13.3159G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/2/48)/grid(1/108/1)/384]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/512/manual_time         19.3 us          100 us          735 bytes_per_second=19.1452G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/3/64)/grid(1/108/1)/768]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/640/manual_time         22.1 us          105 us          635 bytes_per_second=20.8863G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/3/80)/grid(1/108/1)/960]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/768/manual_time         20.7 us          102 us          679 bytes_per_second=26.7507G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/4/96)/grid(1/108/1)/1536]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/896/manual_time         26.7 us          109 us          523 bytes_per_second=24.1186G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/4/112)/grid(1/108/1)/1792]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/1023/manual_time        24.8 us          106 us          565 bytes_per_second=29.7024G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 4 / /Launch_Parameters[block(1/2/256)/grid(1/216/1)/2048]
NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16___GRAPH/NvFuserScheduler_LayerNorm_BWD_MultiReductionsPerBlock_fp16/128/1024/manual_time        24.1 us          106 us          580 bytes_per_second=30.5296G/s Red On Fastest Dim // Persistent Kernel // Project Persistent Buffers //  // Iteration Domain: multiple reductions per block / split grid dimension /  // Inner Reduction Domain: pad to warp / persistent batch - 1 / vectorize / factor 8/Launch_Parameters[block(1/4/128)/grid(1/108/1)/2048]
```